### PR TITLE
Proposed waitress introduction as windows webserver

### PR DESCRIPTION
**Pull request by akshaya-a**
The main purpose of this PR is to introduce mlflow ui functionality on windows. I added on the server CLI command as well, but am unsure of the implications until we get the Windows FileStore support in. 

At the moment I opted not to replace gunicorn in all code paths after both reviewing the gunicorn-related commit history (there appear to have been customer use cases driving some of them) and kept the changes platform-dependent (ie assuming no one is interested in waitress on non-win32 platforms), but am happy to revisit that if there's interest.

Current status:
mlflow ui works inside waitress-serve (pretty much a drop-in replacement). mlflow server needs @eedeleon's windows support change to get filestore working locally.

Requires:
1. ~CLI syntax/signature feedback from interested parties~
2. ~#996 for some basic automated validation~ We've decided to add some basic selenium testing in a followup PR, screenshots attached here for manual validation

Manual validation:
1. Followed dev instructions for npm run build
2. mlflow (pip install -e .) ui/server <- works
3. underlying waitress-serve command <- works

image

### CODE REVIEW

File: mlflow/cli.py
Comment (tomasatdatabricks) [inline mlflow/cli.py]: should we raise exception if we got waitress options similarly to above?

File: mlflow/cli.py
Comment (tomasatdatabricks) [inline mlflow/cli.py]: is static_prefix understood only by gunicorn?

File: mlflow/cli.py
Comment (tomasatdatabricks) [inline mlflow/cli.py]: It looks like the only purpose of this method is to check that gunicorn opts are not passed on windows (and maybe we want to check that waitress is not use on non-windows, at least for now). I think it would be simpler if we do not return anything in that case.

File: mlflow/cli.py
Comment (tomasatdatabricks) [inline mlflow/cli.py]: I think just adding waitress_opts is easier to read than passing in **server_kwargs.

File: mlflow/server/__init__.py
Comment (tomasatdatabricks) [inline mlflow/server/__init__.py]: is this understood by waitress as well?

File: mlflow/server/__init__.py
Comment (tomasatdatabricks) [inline mlflow/server/__init__.py]: post -> port ?

File: mlflow/cli.py
Comment (tomasatdatabricks) [inline mlflow/cli.py]: missing ','?

**Review by tomasatdatabricks — COMMENTED** [review]
Thanks for the contribution! 

Looks good overall, but there seems to be some typos and linter issues + I left some comments in code.

Can you also provide brief description / (testcase?) of how was this tested in windows? We do not have windows tests on Travis yet, but it would be nice to have a way to verify this PR somehow. I can check the PR out on windows machine and run the tests.

File: mlflow/cli.py
Comment (akshaya-a) [reply mlflow/cli.py]: @tomasatdatabricks my limited understanding is that this is a flask routing concern and not the actual webserver (and a quick search around the codebase appears to reflect that) , but I'm certainly out of my depth here - happy to learn how it gets leveraged by gunicorn

static_prefix -> env_map[STATIC_PREFIX_ENV_VAR] -> _add_static_prefix -> @app.route(_add_static_prefix('/get-artifact'))

File: mlflow/cli.py
Comment (akshaya-a) [reply mlflow/cli.py]: Yes, that's a great idea

File: mlflow/cli.py
Comment (tomasatdatabricks) [reply mlflow/cli.py]: I see, that makes sense, thanks for the clarification!

File: mlflow/cli.py
Comment (akshaya-a) [reply mlflow/cli.py]: Switch to a pure validation function, but tweaked the default handling for "workers" to detect a manual selection of the default value - please take a look and let me know what you think!

**Comment by akshaya-a** [comment]
looks like an unrelated pip download network connectivity issue has failed one of the Large builds and I don't think I have privileges to requeue without pushing a new commit

File: mlflow/pyfunc/backend.py
Comment (tomasatdatabricks) [inline mlflow/pyfunc/backend.py]: Nit: this is not necessary.

**Review by tomasatdatabricks — APPROVED** [review]
Thanks for the updates @akshaya-a!

It looks good, I only had one nit-comment about unnecessary changes in pyfunc backend, but I don't think that should block the PR.

I restarted the failed test and I will merge this when it succeeds.

File: setup.py
Comment (mbelang) [inline setup.py]: Why `waitress` only for windows system? Why not let the user choose one or the other for linux systems? I'm not a big fan of gunicorn and I'd like to replace it with an other server. Since it is embedded in mlflow, that would be neat if user could specify his own...

**Pull request merged**
